From f7df95a36940e7864277d3a031a360890e4bfad0 Mon Sep 17 00:00:00 2001
From: Andrei Warkentin <andrei.warkentin@intel.com>
Date: Thu, 16 Mar 2023 15:46:04 -0500
Subject: [PATCH 1/1] RiscVVirt: Patches to enable FdtBusPkg components

Ports over the following components to use FdtBusDxe
and driver binding where possible.

- VirtioFdtDxe
- PciHostBridgeLibEcam
- FdtPciPcdProducerLib
- HighMemDxe (binding variant by default)
- PciSioSerialDxe (which supports the FDT UART device),
  instead of SerialDxe.

The changes to PlatformBm are a bit suboptimal. Need to lookup
the UART by alias or /chosen/stdout-path.

Signed-off-by: Andrei Warkentin <andrei.warkentin@intel.com>
---
 OvmfPkg/RiscVVirt/RiscVVirtQemu.dsc           | 25 ++++--
 OvmfPkg/RiscVVirt/RiscVVirtQemu.fdf           | 11 ++-
 .../PlatformBootManagerLib.inf                |  3 +
 .../PlatformBootManagerLib/PlatformBm.c       | 77 +++++++++++--------
 4 files changed, 75 insertions(+), 41 deletions(-)

diff --git a/OvmfPkg/RiscVVirt/RiscVVirtQemu.dsc b/OvmfPkg/RiscVVirt/RiscVVirtQemu.dsc
index f8b9479345d7..85c3dbfc427d 100644
--- a/OvmfPkg/RiscVVirt/RiscVVirtQemu.dsc
+++ b/OvmfPkg/RiscVVirt/RiscVVirtQemu.dsc
@@ -95,9 +95,10 @@ [LibraryClasses.common]
   FrameBufferBltLib|MdeModulePkg/Library/FrameBufferBltLib/FrameBufferBltLib.inf
   QemuBootOrderLib|OvmfPkg/Library/QemuBootOrderLib/QemuBootOrderLib.inf
   FileExplorerLib|MdeModulePkg/Library/FileExplorerLib/FileExplorerLib.inf
-  PciPcdProducerLib|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+  # PciPcdProducerLib|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+  PciPcdProducerLib|FdtBusPkg/Library/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
   PciSegmentLib|MdePkg/Library/BasePciSegmentLibPci/BasePciSegmentLibPci.inf
-  PciHostBridgeLib|OvmfPkg/Fdt/FdtPciHostBridgeLib/FdtPciHostBridgeLib.inf
+  PciHostBridgeLib|FdtBusPkg/Library/PciHostBridgeLibEcam/PciHostBridgeLibEcam.inf
   PciHostBridgeUtilityLib|OvmfPkg/Library/PciHostBridgeUtilityLib/PciHostBridgeUtilityLib.inf
   PeiHardwareInfoLib|OvmfPkg/Library/HardwareInfoLib/PeiHardwareInfoLib.inf
   PlatformHookLib|MdeModulePkg/Library/BasePlatformHookLibNull/BasePlatformHookLibNull.inf
@@ -334,7 +335,8 @@ [Components]
   MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitterDxe.inf
   MdeModulePkg/Universal/Console/GraphicsConsoleDxe/GraphicsConsoleDxe.inf
   MdeModulePkg/Universal/Console/TerminalDxe/TerminalDxe.inf
-  MdeModulePkg/Universal/SerialDxe/SerialDxe.inf
+  # MdeModulePkg/Universal/SerialDxe/SerialDxe.inf
+  FdtBusPkg/Drivers/PciSioSerialDxe/PciSioSerialDxe.inf
 
   MdeModulePkg/Universal/HiiDatabaseDxe/HiiDatabaseDxe.inf
 
@@ -351,14 +353,18 @@ [Components]
   #
   # Platform Driver
   #
-  OvmfPkg/Fdt/VirtioFdtDxe/VirtioFdtDxe.inf
+  # OvmfPkg/Fdt/VirtioFdtDxe/VirtioFdtDxe.inf
+  FdtBusPkg/Drivers/VirtioFdtDxe/VirtioFdtDxe.inf
+  FdtBusPkg/Drivers/FdtBusDxe/FdtBusDxe.inf
   EmbeddedPkg/Drivers/FdtClientDxe/FdtClientDxe.inf {
     <LibraryClasses>
       DevicePathLib|MdePkg/Library/UefiDevicePathLib/UefiDevicePathLib.inf
       PcdLib|MdePkg/Library/BasePcdLibNull/BasePcdLibNull.inf
       DebugLib|MdePkg/Library/BaseDebugLibNull/BaseDebugLibNull.inf
   }
-  OvmfPkg/Fdt/HighMemDxe/HighMemDxe.inf
+  #OvmfPkg/Fdt/HighMemDxe/HighMemDxe.inf
+  FdtBusPkg/Drivers/HighMemDxe/HighMemDxe.inf
+  #FdtBusPkg/Drivers/HighMemDxe/HighMemDxeNoBinding.inf
   OvmfPkg/VirtioBlkDxe/VirtioBlk.inf
   OvmfPkg/VirtioScsiDxe/VirtioScsi.inf
   OvmfPkg/VirtioNetDxe/VirtioNet.inf
@@ -448,12 +454,14 @@ [Components]
   #
   OvmfPkg/RiscVVirt/PciCpuIo2Dxe/PciCpuIo2Dxe.inf {
     <LibraryClasses>
-      NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      # NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      NULL|FdtBusPkg/Library/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
   }
   MdeModulePkg/Bus/Pci/PciHostBridgeDxe/PciHostBridgeDxe.inf
   MdeModulePkg/Bus/Pci/PciBusDxe/PciBusDxe.inf {
     <LibraryClasses>
-      NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      # NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      NULL|FdtBusPkg/Library/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
   }
   OvmfPkg/PciHotPlugInitDxe/PciHotPlugInit.inf
   OvmfPkg/VirtioPciDeviceDxe/VirtioPciDeviceDxe.inf
@@ -503,5 +511,6 @@ [Components]
   MdeModulePkg/Universal/Acpi/BootGraphicsResourceTableDxe/BootGraphicsResourceTableDxe.inf
   OvmfPkg/AcpiPlatformDxe/AcpiPlatformDxe.inf {
     <LibraryClasses>
-      NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      # NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      NULL|FdtBusPkg/Library/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
   }
diff --git a/OvmfPkg/RiscVVirt/RiscVVirtQemu.fdf b/OvmfPkg/RiscVVirt/RiscVVirtQemu.fdf
index 40d12e0f4c46..792144bfcab6 100644
--- a/OvmfPkg/RiscVVirt/RiscVVirtQemu.fdf
+++ b/OvmfPkg/RiscVVirt/RiscVVirtQemu.fdf
@@ -69,9 +69,13 @@ [FV.DXEFV]
 INF  MdeModulePkg/Core/Dxe/DxeMain.inf
 INF  MdeModulePkg/Universal/PCD/Dxe/Pcd.inf
 INF  MdeModulePkg/Universal/DevicePathDxe/DevicePathDxe.inf
-INF  OvmfPkg/Fdt/VirtioFdtDxe/VirtioFdtDxe.inf
+# INF  OvmfPkg/Fdt/VirtioFdtDxe/VirtioFdtDxe.inf
+INF  FdtBusPkg/Drivers/VirtioFdtDxe/VirtioFdtDxe.inf
+INF  FdtBusPkg/Drivers/FdtBusDxe/FdtBusDxe.inf
 INF  EmbeddedPkg/Drivers/FdtClientDxe/FdtClientDxe.inf
-INF  OvmfPkg/Fdt/HighMemDxe/HighMemDxe.inf
+# INF  OvmfPkg/Fdt/HighMemDxe/HighMemDxe.inf
+INF  FdtBusPkg/Drivers/HighMemDxe/HighMemDxe.inf
+# INF  FdtBusPkg/Drivers/HighMemDxe/HighMemDxeNoBinding.inf
 
 #
 # PI DXE Drivers producing Architectural Protocols (EFI Services)
@@ -98,7 +102,8 @@ [FV.DXEFV]
 INF  MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitterDxe.inf
 INF  MdeModulePkg/Universal/Console/GraphicsConsoleDxe/GraphicsConsoleDxe.inf
 INF  MdeModulePkg/Universal/Console/TerminalDxe/TerminalDxe.inf
-INF  MdeModulePkg/Universal/SerialDxe/SerialDxe.inf
+# INF  MdeModulePkg/Universal/SerialDxe/SerialDxe.inf
+INF  FdtBusPkg/Drivers/PciSioSerialDxe/PciSioSerialDxe.inf
 
 # RISC-V Core Drivers
 INF  UefiCpuPkg/CpuTimerDxeRiscV64/CpuTimerDxeRiscV64.inf
diff --git a/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBootManagerLib.inf b/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBootManagerLib.inf
index 9d66c8110c53..d9dd7391f547 100644
--- a/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBootManagerLib.inf
+++ b/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBootManagerLib.inf
@@ -32,6 +32,7 @@ [Packages]
   OvmfPkg/OvmfPkg.dec
   SecurityPkg/SecurityPkg.dec
   ShellPkg/ShellPkg.dec
+  FdtBusPkg/FdtBusPkg.dec
 
 [LibraryClasses]
   BaseLib
@@ -73,3 +74,5 @@ [Protocols]
   gEfiGraphicsOutputProtocolGuid
   gEfiPciRootBridgeIoProtocolGuid
   gVirtioDeviceProtocolGuid
+  gEfiDtIoProtocolGuid
+  gEfiDevicePathProtocolGuid
diff --git a/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBm.c b/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBm.c
index 44fd21f74fcf..e60a149407f3 100644
--- a/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBm.c
+++ b/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBm.c
@@ -18,6 +18,7 @@
 #include <Library/QemuBootOrderLib.h>
 #include <Library/TpmPlatformHierarchyLib.h>
 #include <Library/UefiBootManagerLib.h>
+#include <Library/DevicePathLib.h>
 #include <Protocol/DevicePath.h>
 #include <Protocol/FirmwareVolume2.h>
 #include <Protocol/GraphicsOutput.h>
@@ -25,6 +26,7 @@
 #include <Protocol/PciIo.h>
 #include <Protocol/PciRootBridgeIo.h>
 #include <Protocol/VirtioDevice.h>
+#include <Protocol/DtIo.h>
 #include <Guid/EventGroup.h>
 #include <Guid/GlobalVariable.h>
 #include <Guid/RootBridgesConnectedEventGroup.h>
@@ -39,27 +41,18 @@
 
 #pragma pack (1)
 typedef struct {
-  VENDOR_DEVICE_PATH            SerialDxe;
   UART_DEVICE_PATH              Uart;
   VENDOR_DEFINED_DEVICE_PATH    TermType;
   EFI_DEVICE_PATH_PROTOCOL      End;
 } PLATFORM_SERIAL_CONSOLE;
 #pragma pack ()
 
-STATIC PLATFORM_SERIAL_CONSOLE  mSerialConsole = {
-  //
-  // VENDOR_DEVICE_PATH SerialDxe
-  //
-  {
-    { HARDWARE_DEVICE_PATH,  HW_VENDOR_DP, DP_NODE_LEN (VENDOR_DEVICE_PATH) },
-    EDKII_SERIAL_PORT_LIB_VENDOR_GUID
-  },
-
+STATIC PLATFORM_SERIAL_CONSOLE  mSerialConsoleSuffix = {
   //
   // UART_DEVICE_PATH Uart
   //
   {
-    { MESSAGING_DEVICE_PATH, MSG_UART_DP,  DP_NODE_LEN (UART_DEVICE_PATH)   },
+    { MESSAGING_DEVICE_PATH, MSG_UART_DP, DP_NODE_LEN (UART_DEVICE_PATH) },
     0,                                      // Reserved
     FixedPcdGet64 (PcdUartDefaultBaudRate), // BaudRate
     FixedPcdGet8 (PcdUartDefaultDataBits),  // DataBits
@@ -767,6 +760,12 @@ PlatformRegisterOptionsAndKeys (
 // BDS Platform Functions
 //
 
+VOID
+EFIAPI
+EfiBootManagerConnectAllConsoles (
+  VOID
+  );
+
 /**
   Do the platform init, can be customized by OEM/IBV
   Possible things that can be done in PlatformBootManagerBeforeConsole:
@@ -784,9 +783,13 @@ PlatformBootManagerBeforeConsole (
   VOID
   )
 {
-  UINT16         FrontPageTimeout;
-  RETURN_STATUS  PcdStatus;
-  EFI_STATUS     Status;
+  UINT16                    FrontPageTimeout;
+  RETURN_STATUS             PcdStatus;
+  EFI_STATUS                Status;
+  EFI_HANDLE                Handle;
+  EFI_DT_IO_PROTOCOL        *DtIo;
+  EFI_DEVICE_PATH_PROTOCOL  *DtDp;
+  EFI_DEVICE_PATH_PROTOCOL  *NewDp;
 
   //
   // Signal EndOfDxe PI Event
@@ -840,23 +843,31 @@ PlatformBootManagerBeforeConsole (
   //
   // Add the hardcoded serial console device path to ConIn, ConOut, ErrOut.
   //
-  CopyGuid (&mSerialConsole.TermType.Guid, &gEfiTtyTermGuid);
+  CopyGuid (&mSerialConsoleSuffix.TermType.Guid, &gEfiTtyTermGuid);
 
-  EfiBootManagerUpdateConsoleVariable (
-    ConIn,
-    (EFI_DEVICE_PATH_PROTOCOL *)&mSerialConsole,
-    NULL
-    );
-  EfiBootManagerUpdateConsoleVariable (
-    ConOut,
-    (EFI_DEVICE_PATH_PROTOCOL *)&mSerialConsole,
-    NULL
-    );
-  EfiBootManagerUpdateConsoleVariable (
-    ErrOut,
-    (EFI_DEVICE_PATH_PROTOCOL *)&mSerialConsole,
-    NULL
-    );
+  Status = gBS->LocateProtocol (
+                  &gEfiDtIoProtocolGuid,
+                  NULL,
+                  (VOID **)&DtIo
+                  );
+  ASSERT_EFI_ERROR (Status);
+
+  Status = DtIo->Lookup (DtIo, "/soc/serial@10000000", TRUE, &Handle);
+  ASSERT_EFI_ERROR (Status);
+
+  Status = gBS->HandleProtocol (
+                  Handle,
+                  &gEfiDevicePathProtocolGuid,
+                  (VOID **)&DtDp
+                  );
+  ASSERT_EFI_ERROR (Status);
+
+  NewDp = AppendDevicePath (DtDp, (VOID *)&mSerialConsoleSuffix);
+
+  EfiBootManagerUpdateConsoleVariable (ConIn, NewDp, NULL);
+  EfiBootManagerUpdateConsoleVariable (ConOut, NewDp, NULL);
+  EfiBootManagerUpdateConsoleVariable (ErrOut, NewDp, NULL);
+  FreePool (NewDp);
 
   //
   // Set the front page timeout from the QEMU configuration.
@@ -954,6 +965,12 @@ PlatformBootManagerAfterConsole (
     // Connect the rest of the devices.
     //
     EfiBootManagerConnectAll ();
+    //
+    // Uncomment if you want to use an alternate console instead of
+    // the /soc/serial@10000000 one.
+    //
+    // EfiBootManagerConnectAllConsoles ();
+    //
   }
 
   //
-- 
2.34.1

